# Q0: import dash standard libraries

In [3]:
from dash import Dash, dcc, html, Input, Output
from jupyter_dash import JupyterDash

# Q1: Make dash that prints "Data Science"

In [1]:
from dash import Dash, html

app = Dash(__name__)

app.layout = html.Div([
    html.H1(children='Data Science')
])

if __name__ == '__main__':
  
    app.run() # app.run_server() also works

Dash is running on http://127.0.0.1:8050/

 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:8050
Press CTRL+C to quit
127.0.0.1 - - [31/May/2023 13:08:15] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 13:08:15] "GET /_dash-layout HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 13:08:15] "GET /_dash-dependencies HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 13:08:15] "GET /_favicon.ico?v=2.10.1 HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 13:08:19] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 13:08:19] "GET /_dash-layout HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 13:08:19] "GET /_dash-dependencies HTTP/1.1" 200 -


# Q2: Make dash that takes name input and outpt "Welcome to Data Science dashboard {name}"

In [1]:
from dash import Dash, dcc, html, Input, Output
from jupyter_dash import JupyterDash

app = JupyterDash(__name__)

app.layout = html.Div([
    html.H2("Welcome to Data Science dashboard"),
    html.Div([
        "Input: ",
        
        dcc.Input(id='my-input', type='text')
    ]),
    html.Br(),
    html.Div(id='my-output'),

])


@app.callback(
    Output(component_id='my-output', component_property='children'),
    Input(component_id='my-input', component_property='value')
)
def update_output_div(a):
    return f'Welcome to Data Science dashboard: {a}'


if __name__ == '__main__':
    app.run_server()

Dash is running on http://127.0.0.1:8050/

Dash app running on http://127.0.0.1:8050/


# Q3: Make Plotly-Dash that takes two inputs (numbers) and output 3 numbers x, y , z= x+y
- hint: use html.Div, dcc.Input(type='number')

In [1]:
from dash import Dash, dcc, html, Input, Output
from jupyter_dash import JupyterDash

app = JupyterDash(__name__)

app.layout = html.Div([
    html.H2("Two Numbers "),
    html.Div([
        "Input: ",
        dcc.Input(id='my-input', value=0,type='number')
    ]),
    html.Div([
        "Input: ",
        dcc.Input(id='my-input2',value=0,type='number')
    ]),
    html.Br(),
    html.Div(id='my-output'),

])


@app.callback(
    Output(component_id='my-output', component_property='children'),

    Input(component_id='my-input', component_property='value'),
    Input(component_id='my-input2', component_property='value')
)
def update_output_div(input_value, input_value2):
    return f'First Number {input_value} and the second: {input_value2} the res {input_value + input_value2}'


if __name__ == '__main__':
    app.run_server()

Dash is running on http://127.0.0.1:8050/

Dash app running on http://127.0.0.1:8050/


# Q4.1: Make JubyterDash with Rangeslider from 1-15 with 6 option

In [1]:
from dash import Dash, dcc, html, Input, Output
import plotly.graph_objects as go
import numpy as np
from jupyter_dash import JupyterDash
app = JupyterDash(__name__)

app.layout = html.Div([
    html.H1('RangeSlider Example'),
    dcc.RangeSlider(
        id='my-slider',
        min=1,
        max=15,
        value=[1],)
        
])



def update_bar_chart(size):
    return size
app.run_server()

Dash is running on http://127.0.0.1:8050/

Dash app running on http://127.0.0.1:8050/


# Q4.2: add to the previous dash: Label, set default value on rangeslider as 7-10

In [1]:
from dash import Dash, dcc, html, Input, Output
import plotly.graph_objects as go
import numpy as np
from jupyter_dash import JupyterDash
app = JupyterDash(__name__)

app.layout = html.Div([
    html.H1('RangeSlider Example'),
    html.Label('Select a range of values:'),
    dcc.RangeSlider(
        id='my-slider',
        min=1,
        max=15,
        value=[7,10],)
        
])



def update_bar_chart(size):
    return size
app.run_server()

Dash is running on http://127.0.0.1:8050/

Dash app running on http://127.0.0.1:8050/


# Q5: Make 3 different Dcc types (Dropdown - Checkbox...) in one dash

In [1]:
from dash import Dash, dcc, html, Input, Output
import plotly.graph_objects as go
import numpy as np
from jupyter_dash import JupyterDash
app = JupyterDash(__name__)
app.layout = html.Div([
    html.H1("Dash Core Components Example"),
    html.Label("Select an option from the dropdown:"),
    dcc.Dropdown(['LA', 'NYC', 'MTL'],
        id="my-dropdown",
    ),
    html.Br(),
    dcc.Checklist(
        ['New York City', 'Montréal', 'San Francisco'],
        id="my-checkbox",
    ),
    html.Br(),
    html.Label("Select one option from the radio items:"),
     dcc.RadioItems(['Linear', 'Log'],
        id="my-radio",
    )
])

if __name__ == '__main__':
    app.run_server(debug=True)


Dash is running on http://127.0.0.1:8050/

Dash app running on http://127.0.0.1:8050/


# Q6: Make Plotly chart, then add it to a plotly_dash

In [1]:
import plotly.express as px
from dash import Dash, dcc, html, Input, Output
import plotly.graph_objects as go
import numpy as np
from jupyter_dash import JupyterDash
df = px.data.gapminder()
fig = px.scatter(df.query("year==2007"), x="gdpPercap", y="lifeExp", size="pop", color="continent",
                 hover_name="country", log_x=True, size_max=60)

# Create a Plotly Dash app
app = Dash(__name__)

app.layout = html.Div([
    html.H1("Plotly Dash Example"),
    dcc.Graph(id="my-graph", figure=fig)
])

if __name__ == '__main__':
    app.run_server()

Dash is running on http://127.0.0.1:8050/

 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:8050
Press CTRL+C to quit
127.0.0.1 - - [31/May/2023 18:57:40] "GET /_reload-hash HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 18:57:40] "GET /_reload-hash HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 18:57:40] "GET /_reload-hash HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 18:57:40] "GET /_reload-hash HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 18:57:40] "GET /_reload-hash HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 18:57:40] "GET /_dash-dependencies HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 18:57:40] "GET /_reload-hash HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 18:57:40] "GET /_reload-hash HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 18:57:40] "GET /_dash-dependencies HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 18:57:40] "GET /_dash-dependencies HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 18:57:40] "GET /_reload-hash HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 18:57:40] "GET /_dash-layout HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 18:57:40] "GET /_reload-hash HTTP/

# Q7: From our data make plotly-dash with suitble chart, and Dropdown.

- Dropdown should contain at least 3 options, each one represents different column.
- Add label to dropdown.
- The chart data should change depending on the selected dropdown option.

In [1]:
import plotly.express as px
from dash import Dash, dcc, html, Input, Output
import plotly.graph_objects as go
import numpy as np
from jupyter_dash import JupyterDash
import pandas as pd
df = pd.read_csv("titanic passenger list.csv")
app = JupyterDash(__name__)

app.layout = html.Div([
    html.Label('Select a Column'),
    dcc.Dropdown(
        id='column-dropdown',
        options=[
            {'label': 'Age', 'value': 'Age'},
            {'label': 'Fare', 'value': 'Fare'},
            {'label': 'Survived', 'value': 'Survived'}
        ],
        value='Age'
    ),
    dcc.Graph(id='chart')
])
@app.callback(
    Output('chart', 'figure'),
    Input('column-dropdown', 'value')
)
def update_chart(column):
    fig = px.histogram(df, x=column.lower())
    return fig
if __name__ == '__main__':
    app.run_server()

Dash is running on http://127.0.0.1:8050/

Dash app running on http://127.0.0.1:8050/


# Q8: add 2 more charts to the prevoius Ploty-Dash

In [1]:
import plotly.express as px
from dash import Dash, dcc, html, Input, Output
import plotly.graph_objects as go
import numpy as np
from jupyter_dash import JupyterDash
import pandas as pd
df = pd.read_csv("titanic passenger list.csv")
app = Dash(__name__)
app.layout = html.Div([
    html.Label('Select a Column'),
    dcc.Dropdown(
        id='column-dropdown',
        options=[
            {'label': 'Age', 'value': 'age'},
            {'label': 'Fare', 'value': 'fare'},
            {'label': 'Survived', 'value': 'survived'}
        ],
        value='age'
    ),
    dcc.Graph(id='histogram'),
    dcc.Graph(id='scatter'),
    dcc.Graph(id='pie')
])
@app.callback(
    Output('histogram', 'figure'),
    Input('column-dropdown', 'value')
)
def update_histogram(column):
    fig = px.histogram(df, x=column.lower())
    return fig

@app.callback(
    Output('scatter', 'figure'),
    Input('column-dropdown', 'value')
)
def update_scatter(column):
    fig = px.scatter(df, x='age', y='fare', color=column.lower())
    return fig

@app.callback(
    Output('pie', 'figure'),
    Input('column-dropdown', 'value')
)
def update_pie(column):
    fig = px.pie(df, names='sex', values=column)
    return fig
if __name__ == '__main__':
    app.run_server()

Dash is running on http://127.0.0.1:8050/

 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:8050
Press CTRL+C to quit
127.0.0.1 - - [31/May/2023 19:14:12] "GET /_reload-hash HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:14:12] "GET /_dash-dependencies HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:14:12] "GET /_dash-layout HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:14:12] "GET /_reload-hash HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:14:12] "GET /_dash-layout HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:14:12] "GET /_dash-dependencies HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:14:13] "POST /_dash-update-component HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:14:13] "POST /_dash-update-component HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:14:13] "POST /_dash-update-component HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:14:13] "POST /_dash-update-component HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:14:13] "POST /_dash-update-component HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:14:13] "POST /_dash-update-component HTTP/1.1" 200 -
1

# Q9: Search for "Plotly Gallery" and fetch for a graph and use it with our data.

- Note: You need to put the resourse addres of fetched your grapth.

In [1]:
import plotly.express as px
from dash import Dash, dcc, html, Input, Output
import plotly.graph_objects as go
import numpy as np
from jupyter_dash import JupyterDash
import pandas as pd


df=pd.read_csv("titanic passenger list.csv")


app = Dash(__name__)

app.layout = html.Div([
    html.H4('Show Survived based on 2 x-axis '),
    dcc.Dropdown(
        id="dropdown",
        options=['embarked','sex'],
        #value=['embarked', 'sex'],
        multi=True
    ),
    dcc.Graph(id="graph"),
])


@app.callback(
    Output("graph", "figure"), 
    Input("dropdown", "value"))
def update_bar_chart(x_axis):
    return px.histogram(df, x=x_axis)


app.run_server()



Dash is running on http://127.0.0.1:8050/

 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:8050
Press CTRL+C to quit
127.0.0.1 - - [31/May/2023 21:13:22] "GET /_reload-hash HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 21:13:23] "GET /_reload-hash HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 21:13:25] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 21:13:26] "GET /_dash-layout HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 21:13:26] "GET /_dash-dependencies HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 21:13:26] "GET /_dash-component-suites/dash/dcc/async-dropdown.js HTTP/1.1" 304 -
127.0.0.1 - - [31/May/2023 21:13:26] "GET /_dash-component-suites/dash/dcc/async-graph.js HTTP/1.1" 304 -
127.0.0.1 - - [31/May/2023 21:13:26] "GET /_reload-hash HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 21:13:26] "GET /_dash-component-suites/dash/dcc/async-plotlyjs.js HTTP/1.1" 304 -
[2023-05-31 21:13:26,774] ERROR in app: Exception on /_dash-update-component [POST]
Traceback (most recent call last):
  File "c:\Users\LENOVO\anaconda3\lib\site-packages\flask\app.py", line 25

# Challenge 1.1: Create Plotly dash with two charts depends on different columns of the data
- First Chart : Pie
- Secound Chart: is by your Choice ( Not Pie )

# Challenge 1.2: Make Hovering over a slice of the pie changes the data for the second chart depending on what the slice is.